File: Windows_csharp/Socks5Message.cs
/*
 * Title:  spider Socks5Message.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Linq;
using System.Net;
using System.Runtime.InteropServices;
using System.Text;

namespace spider
{
    public class Socks5Message : Message
    {
        private const int INET6_ADDR_STRING_LENGTH = 46;

        private uint messageId;
        private uint connectionId;
        private uint clientId;
        private uint serverId;
        private char sourceNodeType;
        private string sourceIp;
        private char destinationNodeType;
        private string destinationIp;
        private int tvSec;
        private int tvUsec;
        private int forwarderTvSec;
        private int forwarderTvUsec;
        private byte receiveFlag;       // received:1
        private byte receiveResult;     // ok:0  ng:1
        private byte commandResult;     // ok:0  ng:1


        public Socks5Message(char messageType,
                             uint messageId,
                             uint connectionId,
                             uint clientId,
                             uint serverId,
                             char sourceNodeType,
                             string sourceIp,
                             char destinationNodeType,
                             string destinationIp,
                             int tvSec,
                             int tvUsec,
                             int forwarderTvSec,
                             int forwarderTvUsec,
                             int dataSize,
                             byte[] data)
        {
            this.messageType = messageType;
            this.receiveFlag = 0;
            this.receiveResult = 0;
            this.commandResult = 0;
            this.messageId = messageId;
            this.connectionId = connectionId;
            this.clientId = clientId;
            this.serverId = serverId;
            this.sourceNodeType = sourceNodeType;
            this.sourceIp = sourceIp;
            this.destinationNodeType = destinationNodeType;
            this.destinationIp = destinationIp;
            this.tvSec = tvSec;
            this.tvUsec = tvUsec;
            this.forwarderTvSec = forwarderTvSec;
            this.forwarderTvUsec = forwarderTvUsec;
            this.dataSize = dataSize;

            if(this.dataSize > 0)
            {
                this.data = new byte[SPIDER_MESSAGE_DATA_MAX_SIZE];
            }

            if(this.dataSize > 0 &&
               this.dataSize <= SPIDER_MESSAGE_DATA_MAX_SIZE)
            {
                for(int i = 0; i < this.dataSize; i++)
                {
                    this.data[i] = data[i];
                }
            }else if(this.dataSize > SPIDER_MESSAGE_DATA_MAX_SIZE)
            {
#if DEBUGPRINT
                Console.WriteLine("[-] socks5 message data size error: {0}",
                                  this.dataSize);
#endif
                this.dataSize = SPIDER_MESSAGE_DATA_MAX_SIZE;
                for(int i = 0; i < SPIDER_MESSAGE_DATA_MAX_SIZE; i++)
                {
                    this.data[i] = data[i];
                }
            }
        }

        public Socks5Message(char messageType,
                             byte receiveFlag,
                             byte receiveResult,
                             byte commandResult,
                             uint messageId,
                             uint connectionId,
                             uint clientId,
                             uint serverId,
                             char sourceNodeType,
                             string sourceIp,
                             char destinationNodeType,
                             string destinationIp)
        {
            this.messageType = messageType;
            this.receiveFlag = receiveFlag;
            this.receiveResult = receiveResult;
            this.commandResult = commandResult;
            this.messageId = messageId;
            this.connectionId = connectionId;
            this.clientId = clientId;
            this.serverId = serverId;
            this.sourceNodeType = sourceNodeType;
            this.sourceIp = sourceIp;
            this.destinationNodeType = destinationNodeType;
            this.destinationIp = destinationIp;
            this.tvSec = 0;
            this.tvUsec = 0;
            this.forwarderTvSec = 0;
            this.forwarderTvUsec = 0;
            this.dataSize = 0;
            this.data = null;
        }

        public Socks5Message(byte[] array)
        {
            byte[] tmp;

            messageType = (char)array[0];
            receiveFlag = array[1];
            receiveResult = array[2];
            commandResult = array[3];
            messageId = (uint)IPAddress.NetworkToHostOrder((int)BitConverter.ToUInt32(array, 4));
            connectionId = (uint)IPAddress.NetworkToHostOrder((int)BitConverter.ToUInt32(array, 8));
            clientId = (uint)IPAddress.NetworkToHostOrder((int)BitConverter.ToUInt32(array, 12));
            serverId = (uint)IPAddress.NetworkToHostOrder((int)BitConverter.ToUInt32(array, 16));
            sourceNodeType = (char)array[20];

            tmp = new byte[INET6_ADDR_STRING_LENGTH + 1];
            for(int i = 0; i < INET6_ADDR_STRING_LENGTH + 1; i++)
            {
                tmp[i] = array[22 + i];
            }
            tmp = tmp.Where(b => b != 0x00).ToArray();
            sourceIp = Encoding.UTF8.GetString(tmp);
            sourceIp = sourceIp.Trim();

            destinationNodeType = (char)array[70];

            tmp = new byte[INET6_ADDR_STRING_LENGTH + 1];
            for(int i = 0; i < INET6_ADDR_STRING_LENGTH + 1; i++)
            {
                tmp[i] = array[72 + i];
            }
            tmp = tmp.Where(b => b != 0x00).ToArray();
            destinationIp = Encoding.UTF8.GetString(tmp);
            destinationIp = destinationIp.Trim();

            tvSec = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 120));
            tvUsec = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 124));
            forwarderTvSec = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 128));
            forwarderTvUsec = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 132));
            dataSize = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 136));

            if(dataSize > 0)
            {
                this.data = new byte[SPIDER_MESSAGE_DATA_MAX_SIZE];
            }

            if(dataSize > 0 &&
               dataSize <= SPIDER_MESSAGE_DATA_MAX_SIZE)
            {
                for(int i = 0; i < dataSize; i++)
                {
                    this.data[i] = array[144 + i];
                }
            }else if(dataSize > SPIDER_MESSAGE_DATA_MAX_SIZE)
            {
                for(int i = 0; i < SPIDER_MESSAGE_DATA_MAX_SIZE; i++)
                {
                    this.data[i] = array[144 + i];
                }
            }
        }

        public uint MessageId
        {
            get { return messageId; }
            set { messageId = value; }
        }

        public byte ReceiveFlag
        {
            get { return receiveFlag; }
            set { receiveFlag = value; }
        }

        public byte ReceiveResult
        {
            get { return receiveResult; }
            set { receiveResult = value; }
        }

        public byte CommandResult
        {
            get { return commandResult; }
            set { commandResult = value; }
        }

        public uint ConnectionId
        {
            get { return connectionId; }
            set { connectionId = value; }
        }

        public uint ClientId
        {
            get { return clientId; }
            set { clientId = value; }
        }

        public uint ServerId
        {
            get { return serverId; }
            set { serverId = value; }
        }

        public char SourceNodeType
        {
            get { return sourceNodeType; }
            set { sourceNodeType = value; }
        }

        public string SourceIp
        {
            get { return sourceIp; }
            set { sourceIp = value; }
        }

        public char DestinationNodeType
        {
            get { return destinationNodeType; }
            set { destinationNodeType = value; }
        }

        public string DestinationIp
        {
            get { return destinationIp; }
            set { destinationIp = value; }
        }

        public int TvSec
        {
            get { return tvSec; }
            set { tvSec = value; }
        }

        public int TvUsec
        {
            get { return tvUsec; }
            set { tvUsec = value; }
        }

        public int ForwarderTvSec
        {
            get { return forwarderTvSec; }
            set { forwarderTvSec = value; }
        }

        public int ForwarderTvUsec
        {
            get { return forwarderTvUsec; }
            set { forwarderTvUsec = value; }
        }

        public ushort NetworkToHostOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public ushort HostToNetworkOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public override int CopyToBuffer(ref byte[] buffer)
        {
            int length = 0;
            byte[] tmp;

            buffer[0] = (byte)messageType;
            buffer[1] = (byte)receiveFlag;
            buffer[2] = (byte)receiveResult;
            buffer[3] = (byte)commandResult;
            BitConverter.GetBytes((uint)IPAddress.HostToNetworkOrder((int)this.messageId)).CopyTo(buffer, 4);
            BitConverter.GetBytes((uint)IPAddress.HostToNetworkOrder((int)this.connectionId)).CopyTo(buffer, 8);
            BitConverter.GetBytes((uint)IPAddress.HostToNetworkOrder((int)this.clientId)).CopyTo(buffer, 12);
            BitConverter.GetBytes((uint)IPAddress.HostToNetworkOrder((int)this.serverId)).CopyTo(buffer, 16);
            buffer[20] = (byte)sourceNodeType;
            tmp = Encoding.UTF8.GetBytes(sourceIp);
            for(int i = 0; i < tmp.Length; i++)
            {
                buffer[22 + i] = tmp[i];
            }
            buffer[70] = (byte)destinationNodeType;
            tmp = Encoding.UTF8.GetBytes(destinationIp);
            for(int i = 0; i < tmp.Length; i++)
            {
                buffer[72 + i] = tmp[i];
            }
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.tvSec)).CopyTo(buffer, 120);
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.tvUsec)).CopyTo(buffer, 124);
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.forwarderTvSec)).CopyTo(buffer, 128);
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.forwarderTvUsec)).CopyTo(buffer, 132);
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.dataSize)).CopyTo(buffer, 136);
            for(int i = 0; i < this.dataSize; i++)
            {
                buffer[144 + i] = this.data[i];
            }

            length = SpiderMessageHeader.SPIDER_MESSAGE_HEADER_SIZE
                   + this.dataSize;

            return length;
        }
    }
}

File: extra/Windows/socks5message.cpp
/*
 * Title:  spider socks5message.cpp (Windows)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "socks5message.hpp"


namespace spider
{
    Socks5message::Socks5message(char message_type,
                                 uint32_t message_id,
                                 uint32_t connection_id,
                                 uint32_t client_id,
                                 uint32_t server_id,
                                 char source_node_type,
                                 std::string source_ip,
                                 char destination_node_type,
                                 std::string destination_ip,
                                 int32_t tv_sec,
                                 int32_t tv_usec,
                                 int32_t forwarder_tv_sec,
                                 int32_t forwarder_tv_usec,
                                 int32_t data_size,
                                 char *data)
    {
        this->message_type = message_type;
        this->receive_flag = 0;
        this->receive_result = 0;
        this->command_result = 0;
        this->message_id = message_id;
        this->connection_id = connection_id;
        this->client_id = client_id;
        this->server_id = server_id;
        this->source_node_type = source_node_type;
        this->source_ip = source_ip;
        this->destination_node_type = destination_node_type;
        this->destination_ip = destination_ip;
        this->tv_sec = tv_sec;
        this->tv_usec = tv_usec;
        this->forwarder_tv_sec = forwarder_tv_sec;
        this->forwarder_tv_usec = forwarder_tv_usec;
        this->data_size = data_size;

        if(this->data_size > 0)
        {
            this->data = (char *)calloc(SPIDER_MESSAGE_DATA_MAX_SIZE,
                                        sizeof(char));
        }

        if(this->data_size > 0 &&
           this->data_size <= SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
            std::memcpy(this->data,
                        data,
                        this->data_size);
        }else if(this->data_size > SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
#ifdef DEBUGPRINT
            std::printf("[-] socks5 message data size error: %d\n",
                        this->data_size);
#endif
            this->data_size = SPIDER_MESSAGE_DATA_MAX_SIZE;
            std::memcpy(this->data,
                        data,
                        SPIDER_MESSAGE_DATA_MAX_SIZE);
        }
    }

    Socks5message::Socks5message(char message_type,
                                 uint8_t receive_flag,
                                 uint8_t receive_result,
                                 uint8_t command_result,
                                 uint32_t message_id,
                                 uint32_t connection_id,
                                 uint32_t client_id,
                                 uint32_t server_id,
                                 char source_node_type,
                                 std::string source_ip,
                                 char destination_node_type,
                                 std::string destination_ip)
    {
        this->message_type = message_type;
        this->receive_flag = receive_flag;
        this->receive_result = receive_result;
        this->command_result = command_result;
        this->message_id = message_id;
        this->connection_id = connection_id;
        this->client_id = client_id;
        this->server_id = server_id;
        this->source_node_type = source_node_type;
        this->source_ip = source_ip;
        this->destination_node_type = destination_node_type;
        this->destination_ip = destination_ip;
        this->tv_sec = 0;
        this->tv_usec = 0;
        this->forwarder_tv_sec = 0;
        this->forwarder_tv_usec = 0;
        this->data_size = 0;
        this->data = NULL;
    }

    Socks5message::Socks5message(struct spider_message *spider_message)
    {
        this->message_type = spider_message->header.message_type;
        this->receive_flag = spider_message->header.receive_flag;
        this->receive_result = spider_message->header.receive_result;
        this->command_result = spider_message->header.command_result;
        this->message_id = ntohl(spider_message->header.message_id);
        this->connection_id = ntohl(spider_message->header.connection_id);
        this->client_id = ntohl(spider_message->header.client_id);
        this->server_id = ntohl(spider_message->header.server_id);
        this->source_node_type = spider_message->header.source_node_type;
        this->source_ip = spider_message->header.source_ip;
        this->destination_node_type = spider_message->header.destination_node_type;
        this->destination_ip = spider_message->header.destination_ip;
        this->tv_sec = ntohl(spider_message->header.tv_sec);
        this->tv_usec = ntohl(spider_message->header.tv_usec);
        this->forwarder_tv_sec = ntohl(spider_message->header.forwarder_tv_sec);
        this->forwarder_tv_usec = ntohl(spider_message->header.forwarder_tv_usec);
        this->data_size = ntohl(spider_message->header.data_size);

        if(this->data_size > 0)
        {
            this->data = (char *)calloc(SPIDER_MESSAGE_DATA_MAX_SIZE,
                                        sizeof(char));
        }

        if(this->data_size > 0 &&
           this->data_size <= SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
            std::memcpy(this->data,
                        spider_message->data,
                        this->data_size);
        }else if(this->data_size > SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
#ifdef DEBUGPRINT
            std::printf("[-] socks5 message data size error: %d\n",
                        this->data_size);
#endif
            this->data_size = SPIDER_MESSAGE_DATA_MAX_SIZE;
            std::memcpy(this->data,
                        spider_message->data,
                        SPIDER_MESSAGE_DATA_MAX_SIZE);
        }
    }

    Socks5message::~Socks5message()
    {
        if(data != NULL)
        {
            free(data);
        }
    }

    void Socks5message::set_message_id(uint32_t message_id)
    {
        this->message_id = message_id;
    }

    uint32_t Socks5message::get_message_id()
    {
        return message_id;
    }

    void Socks5message::set_connection_id(uint32_t connection_id)
    {
        this->connection_id = connection_id;
    }

    uint32_t Socks5message::get_connection_id()
    {
        return connection_id;
    }

    void Socks5message::set_client_id(uint32_t client_id)
    {
        this->client_id = client_id;
    }

    uint32_t Socks5message::get_client_id()
    {
        return client_id;
    }

    void Socks5message::set_server_id(uint32_t server_id)
    {
        this->server_id = server_id;
    }

    uint32_t Socks5message::get_server_id()
    {
        return server_id;
    }

    void Socks5message::set_destination_node_type(char destination_node_type)
    {
        this->destination_node_type = destination_node_type;
    }

    char Socks5message::get_destination_node_type()
    {
        return destination_node_type;
    }

    void Socks5message::set_destination_ip(std::string destination_ip)
    {
        this->destination_ip = destination_ip;
    }

    std::string Socks5message::get_destination_ip()
    {
        return destination_ip;
    }

    void Socks5message::set_source_node_type(char source_node_type)
    {
        this->source_node_type = source_node_type;
    }

    char Socks5message::get_source_node_type()
    {
        return source_node_type;
    }

    void Socks5message::set_source_ip(std::string source_ip)
    {
        this->source_ip = source_ip;
    }

    std::string Socks5message::get_source_ip()
    {
        return source_ip;
    }

    void Socks5message::set_tv_sec(int32_t tv_sec)
    {
        this->tv_sec = tv_sec;
    }

    int32_t Socks5message::get_tv_sec()
    {
        return tv_sec;
    }

    void Socks5message::set_tv_usec(int32_t tv_usec)
    {
        this->tv_usec = tv_usec;
    }

    int32_t Socks5message::get_tv_usec()
    {
        return tv_usec;
    }

    void Socks5message::set_forwarder_tv_sec(int32_t forwarder_tv_sec)
    {
        this->forwarder_tv_sec = forwarder_tv_sec;
    }

    int32_t Socks5message::get_forwarder_tv_sec()
    {
        return forwarder_tv_sec;
    }

    void Socks5message::set_forwarder_tv_usec(int32_t forwarder_tv_usec)
    {
        this->forwarder_tv_usec = forwarder_tv_usec;
    }

    int32_t Socks5message::get_forwarder_tv_usec()
    {
        return forwarder_tv_usec;
    }

    void Socks5message::set_receive_flag(uint8_t receive_flag)
    {
        this->receive_flag = receive_flag;
    }

    uint8_t Socks5message::get_receive_flag()
    {
        return receive_flag;
    }

    void Socks5message::set_receive_result(uint8_t receive_result)
    {
        this->receive_result = receive_result;
    }

    uint8_t Socks5message::get_receive_result()
    {
        return receive_result;
    }

    void Socks5message::set_command_result(uint8_t command_result)
    {
        this->command_result = command_result;
    }

    uint8_t Socks5message::get_command_result()
    {
        return command_result;
    }

    int32_t Socks5message::copy_to_buffer(char *buffer)
    {
        int32_t length = 0;

        struct spider_message *spider_message = (struct spider_message *)buffer;
        spider_message->header.message_type = this->message_type;
        spider_message->header.receive_flag = this->receive_flag;
        spider_message->header.receive_result = this->receive_result;
        spider_message->header.command_result = this->command_result;
        spider_message->header.message_id = htonl(this->message_id);
        spider_message->header.connection_id = htonl(this->connection_id);
        spider_message->header.client_id = htonl(this->client_id);
        spider_message->header.server_id = htonl(this->server_id);
        spider_message->header.source_node_type = this->source_node_type;
        std::memcpy(spider_message->header.source_ip,
                    this->source_ip.c_str(),
                    this->source_ip.size());
        spider_message->header.destination_node_type = this->destination_node_type;
        std::memcpy(spider_message->header.destination_ip,
                    this->destination_ip.c_str(),
                    this->destination_ip.size());
        spider_message->header.tv_sec = htonl(this->tv_sec);
        spider_message->header.tv_usec = htonl(this->tv_usec);
        spider_message->header.forwarder_tv_sec = htonl(this->forwarder_tv_sec);
        spider_message->header.forwarder_tv_usec = htonl(this->forwarder_tv_usec);
        spider_message->header.data_size = htonl(this->data_size);

        if(this->data_size > 0)
        {
            std::memcpy(spider_message->data,
                        this->data,
                        this->data_size);
        }

        length = sizeof(struct spider_message_header)
               + this->data_size;

        return length;
    }
}

